simulator/controller/controller.go

```diff
@@ -0,0 +1,154 @@
+
+func RunController(client clientset.Interface, cfg *restclient.Config) (func(), error) {
+	controllerCtx, err := createControllerContext(client, cfg, context.Background().Done())
+	if err != nil {
+		return nil, xerrors.Errorf("error building controller context: %v", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	go run(controllerCtx, ctx.Done())
+	shutdownFunc := func() {
+		cancel()
+	}
+
+	return shutdownFunc, nil
+}
+
+// run runs the KubeControllerManagerOptions.  This should never exit.
+func run(controllerCtx controllerContext, stopCh <-chan struct{}) {
+	controllerInitializers := newControllerInitializers()
+	if err := startControllers(context.Background(), controllerCtx, controllerInitializers); err != nil {
+		klog.Fatalf("error starting controllers: %v", err)
+	}
+	controllerCtx.InformerFactory.Start(stopCh)
+	controllerCtx.ObjectOrMetadataInformerFactory.Start(stopCh)
+
+	close(controllerCtx.InformersStarted)
+}
+
+// startControllers starts a set of controllers with a specified controllerContext.
+func startControllers(ctx context.Context, contrllerCtx controllerContext, controllers map[string]initFunc) error {
+	for controllerName, initFn := range controllers {
+		time.Sleep(wait.Jitter(contrllerCtx.ComponentConfig.Generic.ControllerStartInterval.Duration, controllerStartJitter))
+
+		klog.Infof("Starting %q", controllerName)
+		err := initFn(ctx, contrllerCtx)
+		if err != nil {
+			klog.Errorf("Error starting %q", controllerName)
+			return xerrors.Errorf("starting %v: %w", controllerName, err)
+		}
+		klog.Infof("Started %q", controllerName)
+	}
+	return nil
+}
+
+// newControllerInitializers is a public map of named controller groups paired to their initFunc.
+// This allows for structured downstream composition and subdivision.
+func newControllerInitializers() map[string]initFunc {
+	controllers := map[string]initFunc{}
+	controllers["deployment"] = startDeploymentController
+	controllers["replicaset"] = startReplicaSetController
+	controllers["persistent-volume"] = startPersistentVolumeController
+	return controllers
+}
+
+// controllerContext defines the context object for controller.
+
+type controllerContext struct {
+	// ClientBuilder will provide a client for this controller to use
+	ClientBuilder clientbuilder.ControllerClientBuilder
+
+	ComponentConfig kubectrlmgrconfig.KubeControllerManagerConfiguration
+
+	// InformerFactory gives access to informers for the controller.
+	InformerFactory informers.SharedInformerFactory
+
+	// ObjectOrMetadataInformerFactory gives access to informers for typed resources
+	// and dynamic resources by their metadata. All generic controllers currently use
+	// object metadata - if a future controller needs access to the full object this
+	// would become GenericInformerFactory and take a dynamic client.
+	ObjectOrMetadataInformerFactory informerfactory.InformerFactory
+
+	// Stop is the stop channel
+	Stop <-chan struct{}
```

Comment:
Is it needed? Seems not used because we pass `stopch` in `run` function.

Reply:
Ah, you're right! 
I implemented this using [kube-controller-manager](https://github.com/kubernetes/kubernetes/tree/release-1.22/cmd/kube-controller-manager) as an example. 
And the version of k8s used in this project has recently been updated.
I should have removed that line on the previous modification

Reply:
(The channel on the kube-controller-manager appears to have been removed between v1.22 → v1.24.)